### PR TITLE
JITX-4097: Switch v1 queries over to parts-db

### PR DESCRIPTION
**Pull request by ptuckey**
This migrates queries off the V1 DB _almost_ entirely (except STM MCUs).

## Procedural Insertion
The V1 parts have their pins/symbols/landpatterns procedurally inserted by the client.

This continues to happen for the V1 parts (missing landpattern), until we move that logic offline as part of JITX-4198.

## Filtering V1 API to V1 parts
We are still evaluating the quality of the data in the extended parts-db.

In the short-term, we limit the V1 API calls (like `chip-resistor()`) to use the previous V1 parts.

This can be overriden by toggling this design-vars variable:
- DEFAULT-ALLOW-ALL-VENDOR-PART-NUMBERS

We plan to remove this as a follow-up in JITX-4302.

## Test Plan
```
import ocdb/utils/generic-components
view-design-explorer $ chip-resistor(1000.0)
view-design-explorer $ ceramic-cap([])
view-design-explorer $ smd-inductor([])
```

Set `DEFAULT-ALLOW-ALL-VENDOR-PART-NUMBERS` to true, and verify other part:
```
import ocdb/utils/generic-components
clear-dbquery-cache()
view-design-explorer $ chip-resistor(["resistance" => 1300.0, "vendor_part_numbers.lcsc" => "C1471"])
```

### CODE REVIEW

File: utils/parts.stanza [hunk 1]
```diff
@@ -30,7 +30,7 @@ public defn PartsDBComponent (user-properties:Tuple<KeyValue>) -> Component :
     ["_type" => "parts-db"],
     user-properties
   ]
-  parse-component $ dbquery-first(query-properties) as JObject
+  parse-component $ dbquery-first-allow-non-stock(query-properties) as JObject
```
Comment (ptuckey) [inline utils/parts.stanza:33]: Unrelated change - just realized it was wrong while testing...

**Comment by ptuckey** [comment]
This test is failing atm:
```
jstanza run-test ocdb/tests/stm32-mcu-props -named test-mcu-props-in-design
```

That test actually sends a query out to Mongo (via the backend) - and it's failing because we're querying for a capacitor that doesn't exist in parts-db.

There are 2 reasons:
1. Not all the V1 components imported correctly (investigating in JITX-4221)
  - The result is that we don't have a capacitor with the [design-vars/MIN-CAP-VOLTAGE](https://github.com/JITx-Inc/open-components-database/blob/7840bbd8693f25418efb2d6b388bcaaf50a7cd81/utils/design-vars.stanza#L23) of 10.
2. The query times out
  - Even after hacking the above value to `5` (for which there is a result), there are no results
  - This looks to me like a 10-second timeout
  - Need to set up indexes on the parts-db to speed this up

**Comment by ptuckey** [comment]
Creating a MongoDB index for `category` resolved the timeout.

The remaining issues were data-related:
- [design-vars/MIN-CAP-VOLTAGE](https://github.com/JITx-Inc/open-components-database/blob/7840bbd8693f25418efb2d6b388bcaaf50a7cd81/utils/design-vars.stanza#L23)
- [generic-components/smd-query-properties/minimum_quantity](https://github.com/JITx-Inc/open-components-database/blob/17a5d410bd9fda4a08340c3d3296180039bb2907/utils/generic-components.stanza#L168)

Options:
1. Update unit tests to use values that we have results for
2. Fix missing imported v1 components, so that we _do_ have results (JITX-4221)

**Review by PhilippeFerreiraDeSousa — APPROVED** [review]

**Review by erwin-lau — APPROVED** [review]

File: utils/db-parts.stanza [hunk 2]
```diff
@@ -994,4 +1005,12 @@ public defn dbquery-first-allow-non-stock (args: Tuple<KeyValue<String, Tuple<St
...
-
+; Default to a subset of parts while we're vetting an expanded set.
+; TODO: JITX-4302 - Remove this, once we've built more confidnce in the expanded set.
+defn maybe-insert-default-vendor-part-number (exist:Tuple<String>) -> Tuple<String> :
+  if DEFAULT-ALLOW-ALL-VENDOR-PART-NUMBERS :
+    exist
+  else :
+    val exist-vector = to-vector<String> $ exist
+    add(exist-vector, "vendor_part_numbers.digi-key")
+    to-tuple $ exist-vector
```
[the remark this answers is not in the document]
Comment (ptuckey) [reply utils/db-parts.stanza:1016]: Right on, thanks!

File: utils/design-vars.stanza [hunk 3]
```diff
@@ -21,6 +21,7 @@ public var MAX-Z = 500.0
 public var MIN-PKG = "0402"
 public var PREFERRED-MOUNTING = "smd" ; values in ["smd", "through-hole"]
 public var MIN-CAP-VOLTAGE = 10.0 ; Minimum voltage to allow for capacitors
+public var DEFAULT-ALLOW-ALL-VENDOR-PART-NUMBERS:True|False = false
```
[the remark this answers is not in the document]
Comment (ptuckey) [reply utils/design-vars.stanza:24]: Yeah, I think a config would be nicer long-term.

For now, this seemed like a place where users are already overriding config.

This came from Bhusan's request during discussion, that users be able to override this setting.

I'm hopeful we can remove this altogether once we've vetted the parts a little more.

@bhusang: Does this meet your ask? If so, is this a reasonable place for now?

File: components/changjiang-electronics-tech/SS8050.stanza [hunk 4]
```diff
@@ -13,7 +13,6 @@ defpackage ocdb/components/changjiang-electronics-tech/SS8050 :
   import ocdb/utils/symbols
 
   import ocdb/utils/property-structs
-  import ocdb/utils/generator-utils
```
Comment (ptuckey) [inline components/changjiang-electronics-tech/SS8050.stanza:16]: This was a duplicate import (from a few lines up).

It started failing CI after updating to latest jitx-client docker image - maybe new logic to prevent these?

File: utils/parts.stanza [hunk 5]
```diff
@@ -65,20 +65,15 @@ with :
...
-      ; If the landpattern is included, this part should have everything it needs.
+  match(json["category"]):
+    (c:String):
+      switch(c):
+        "capacitor": Capacitor(json)
+        "inductor": Inductor(json)
+        "resistor": Resistor(json)
+        else: Part(json)
+    (c:JNull):
       Part(json)
```
Comment (emspin) [inline utils/parts.stanza:76]: nice cleanup here!

**Review by emspin — APPROVED** [review]

File: utils/design-vars.stanza (hunk 3 above)
[the remark this answers is not in the document]
Comment (PhilippeFerreiraDeSousa) [reply utils/design-vars.stanza:24]: Does Cayden have time to document design-vars.stanza on docs.jitx.com?

File: utils/parts.stanza [hunk 6]
```diff
@@ -65,20 +65,15 @@ with :
...
-    (lp:JObject):
-      ; If the landpattern is included, this part should have everything it needs.
+  match(json["category"]):
+    (c:String):
+      switch(c):
+        "capacitor": Capacitor(json)
+        "inductor": Inductor(json)
+        "resistor": Resistor(json)
+        else: Part(json)
+    (c:JNull):
```
Comment (PhilippeFerreiraDeSousa) [inline utils/parts.stanza:75]: Is there a case it can be null?

File: utils/db-parts.stanza [hunk 7]
```diff
@@ -85,7 +85,7 @@ public defstruct Resistor <: Component :
   TCR: TCR|False ; Temperature coefficient of resistance (ohms/ohm*degC)
   sellers: Tuple<Seller>|False with: (as-method => true)
   resolved-price: Double|False with: (as-method => true)
-  component: ComponentCode|False
```
[the remark this answers is not in the document]
Comment (PhilippeFerreiraDeSousa) [reply utils/db-parts.stanza:88]: Ok, I read the explanation in comment later on

**Review by PhilippeFerreiraDeSousa — APPROVED** [review]
Nice, I am wondering what is in the`ComponentCode` of those v1 parts. I assume you are currently reconstructing what should be in it and it is a work in progress?

File: utils/parts.stanza (hunk 6 above)
Comment (ptuckey) [reply utils/parts.stanza:75]: Yes, if it's a part we can't identify in our "top 10", we leave it empty.

In theory, we could expand the category list so everything has a value - but for now, we only handle the common cases.

File: utils/db-parts.stanza (hunk 7 above)
[the remark this answers is not in the document]
Comment (ptuckey) [reply utils/db-parts.stanza:88]: > Nice, I am wondering what is in the`ComponentCode` of those v1 parts. I assume you are currently reconstructing what should be in it and it is a work in progress?

I think you already saw this, but just in case others are wondering...

For V1 parts, we set the `ComponentCode`, but some fields are left empty (e.g. pins/symbols/landpattern).

We conditionally then insert those fields in the OCDB code - for example:
- [defmethod to-jitx (resistor: Resistor) -> Instantiable :](https://github.com/JITx-Inc/open-components-database/blob/e5cfcb390a5c7d5b6325d23615575d30b272b1b4/utils/db-parts.stanza#L151)

There's a task to move this logic offline: JITX-4198

But, we're being careful about moving that over, as we want to make sure it's done correctly (e.g. "spice" isn't currently part of the parts-db schema).

So in the short-term, we'll stick with the conditional client logic - but, the parts are at least sharing the same backend DB.

[the remark this answers is not in the document]
Comment (PhilippeFerreiraDeSousa) [reply utils/db-parts.stanza:88]: > We conditionally then insert those fields in the OCDB code

My observation is that for v1, we are not using any data in `ComponentCode` yet. So I assume other changes are coming to have `to-jitx` use this data

File: utils/design-vars.stanza (hunk 3 above)
[the remark this answers is not in the document]
Comment (bhusang) [reply utils/design-vars.stanza:24]: @ptuckey This would be ok with me to allow full search over-ride. The design-vars does need documentation, though.

File: utils/db-parts.stanza (hunk 7 above)
[the remark this answers is not in the document]
Comment (ptuckey) [reply utils/db-parts.stanza:88]: Ah, yes - the work for JITX-4198 would include removing the fallback paths entirely - correct!

File: utils/design-vars.stanza (hunk 3 above)
[the remark this answers is not in the document]
Comment (ptuckey) [reply utils/design-vars.stanza:24]: Great!

I see that we already have some docs around these global vars:
- https://docs-testing.jitx.com/reference/utilities/design-vars.html?highlight=design-vars#global-variables

I'll add a note there on this new one (though we hope to remove the need for it, before too long).

[the remark this answers is not in the document]
Comment (ptuckey) [reply utils/design-vars.stanza:24]: Oh right, the docs are a jitx-client change....   I'll do as a follow-up there.

[the remark this answers is not in the document]
Comment (bhusang) [reply utils/design-vars.stanza:24]: Sounds good!
